asn1/CommonLedgerProtocol.asn

```diff
@@ -0,0 +1,103 @@
+CommonLedgerProtocol
+DEFINITIONS
+AUTOMATIC TAGS ::=
+BEGIN
+
+IMPORTS
+    UInt8,
+    UInt32,
+    UInt64,
+    UInt128,
+    UInt256
+    FROM GenericTypes
+
+    Timestamp
+    FROM InterledgerTypes
+
+    InterledgerProtocolError
+    FROM InterledgerProtocol
+;
+
+-- Zero or more protocolData entries, each with a unique
+-- protocolId, describe protocol-specific information about
+-- unpaid, paid, and conditionally paid requests and results
+-- sent over the CommonLedgerProtocol.
+-- The protocolId of each entry describes the protocol for
+-- interpreting the data. Peers need to establish beforehand
+-- which protocols each understand, and which numbers they
+-- use, which can be agreed ad-hoc,except that protocolId 1
+-- is reserved for the Interledger protocol.
+-- When multiple protocolData items are sent in a request,
+-- only the first one will be decisive in triggering the
+-- response, the rest of the protocolData items are considered
+-- side protocols that piggyback on the first one.
+protocolData ::= SEQUENCE OF SEQUENCE {
+  protocolId UInt8 UNIQUE,
```

Comment:
I *much* prefer a string here. You're hardly losing any bytes. `ilp` for example is four bytes vs one byte. And you are so much less likely to have collisions between different vendor-specific extensions and you can see which protocol you're dealing with.

I would also include a mime type, which would be a one byte field with the options 0 = `application/octet-string`, 1 = `text/plain; charset=utf-8`, 2 = `application/json`, with more potentially being added in the future. As I've said in another thread, not including the mime type turned out to be a huge mistake in Ripple Memos, because it made it impossible to print any useful information for debugging.

Maybe one difference in our viewpoints is that I definitely expect more than 256 protocols here. Ripple is going to define its own and I'm sure most vendors will. Most settlement methods will define their own protocols here.